subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java

```diff
@@ -232,4 +274,43 @@ public Configuration getCompileDependencies() {
             return convention.getProject().getConfigurations().getByName(JavaPlugin.COMPILE_CONFIGURATION_NAME);
         }
     }
+
+    /**
+     * A custom artifact type which allows the getFile call to be done lazily only when the
+     * artifact is actually needed.
+     */
+    private abstract static class IntermediateJavaArtifact extends AbstractPublishArtifact {
```

Comment:
I think this class is a sign that we are missing a feature. We tell users not to implement `PublishArtifact`, but then have special implementations ourselves. I think our notation parser for artifacts should understand a `Closure<File>`.